Add initial RISC-V virt board and wire interrupts to the PLIC

michaeljclark opened the pull request:
Closes riscv/riscv-qemu#74

- Add initial cut of the virt board with virtio_block
- Define macros for CPU extension and priv version combinations
- Add U bit to CPUs with user-mode and update mode combinations
- Rename claim_base to context_base in PLIC configuration
- Add enabled_stride and context_stride for PLIC address decoding
- Add PLIC debugging and fix bugs in PLIC address decoding
- Fix PLIC context (hart/mode) parsing in parse_hart_config
- Add irqs to the PLIC during and implement sifive_plic_irq_request
- Change PLIC num_sources to 63 for all boards
- Remove unnecessary check on the irq number in the CLINT
- Rename SiFiveMemmapEntry to MemmapEntry and make memmap static
- Use memmap for all DRAM, ROM and device-tree addresses
- Remove serial naming from spike boards as they don't use serial
- Change sifive_u500 PLIC hart config to contain both M and S mode
- Make RISC-V HTIF and ELF headers locations consistent
- Check RISC-V ELF header magic and make function names uniform
- Fix ELF object leak by calliing elf_close in HTIF driver
- Amend descriptions of the spike boards

To experiment with virtio using the riscv-linux riscv-linux-4.14
branch you need the latest version of bbl with device-tree support
and the following linux-kernel config options enabled:

    CONFIG_RISCV_PLIC=y
    CONFIG_VIRTIO_MMIO=y
    CONFIG_VIRTIO_BLK=y
    CONFIG_CMDLINE="root=/dev/vda ro"

You can then invoke RISC-V QEMU with a virtual disk attached:

    qemu-system-riscv64 -nographic -machine virt \
        -kernel bbl-1.10-virtio \
        -drive file=root.bin,format=raw,id=hd0 \
        -device virtio-blk-device,drive=hd0

michaeljclark commented:
I've run regression tests against the HiFive1 binaries, bbl-1.9, bbl-1.10 and now we have device-tree based virtio-block working. This pull adds ELF verification. e.g. Closes #74

I still have more work to do on VirtIO however this pull makes the PLIC and VirtIO usable with current upstream riscv-linux

```
sh-4.3# cat /proc/interrupts 
           CPU0       
  1:     100000  riscv,plic0,c000000   1  virtio0
sh-4.3# cat /proc/partitions 
major minor  #blocks  name

 254        0     262144 vda
sh-4.3# uname -a
Linux ucbvax 4.14.0-00029-g1c6baf1066cc-dirty #27 Sat Dec 9 19:53:20 NZDT 2017 riscv64 GNU/Linux
```

davidlt commented:
first and foremost, thanks for the work! Highly appreciated.

Does network (e.g. virtio-net-pci) work in guest?

michaeljclark commented:
Not yet. This is a progress commit. It's based on virtio-mmio and exposes virtio-block in device-tree. However if you look at `hw/riscv/virt.c` it should be pretty easy to add. Working on virtio-net next...

I just thought I'd submit a pull because this branch is in a better state than master with respect to the PLIC. virtio-block is the first driver we've used to test the PLIC implementation. I should be able to add virtio-net, virtio-console, etc in the next few days...

michaeljclark commented:
The virt board currently requires SBI console, until we've added virtio-console

michaeljclark commented:
Travis is reporting a format string issue in `create_fdt`... I'll need to patch this branch before the pull request is accepted.

```
/home/travis/build/riscv/riscv-qemu/hw/riscv/sifive_u500.c: In function ‘create_fdt’:
/home/travis/build/riscv/riscv-qemu/hw/riscv/sifive_u500.c:111:5: error: format ‘%llx’ expects argument of type ‘long long unsigned int’, but argument 2 has type ‘hwaddr’ [-Werror=format=]
     nodename = g_strdup_printf("/soc/clint@%llx", memmap[SIFIVE_U500_CLINT].base);
     ^
/home/travis/build/riscv/riscv-qemu/hw/riscv/sifive_u500.c:120:5: error: format ‘%llx’ expects argument of type ‘long long unsigned int’, but argument 2 has type ‘hwaddr’ [-Werror=format=]
     nodename = g_strdup_printf("/memory@%llx", memmap[SIFIVE_U500_DRAM].base);
     ^
/home/travis/build/riscv/riscv-qemu/hw/riscv/sifive_u500.c:168:9: error: format ‘%llx’ expects argument of type ‘long long unsigned int’, but argument 2 has type ‘hwaddr’ [-Werror=format=]
         memmap[SIFIVE_U500_PLIC].base);
         ^
/home/travis/build/riscv/riscv-qemu/hw/riscv/sifive_u500.c:187:5: error: format ‘%llx’ expects argument of type ‘long long unsigned int’, but argument 2 has type ‘hwaddr’ [-Werror=format=]
     nodename = g_strdup_printf("/uart@%llx", memmap[SIFIVE_U500_UART0].base);
     ^
cc1: all warnings being treated as errors
```

davidlt commented:
+1 also for virtio-console. Again thanks for the work!

michaeljclark commented:
Okay. I've changed the device tree entries to use generic `virtio_mmio` nodes and there are 8 transports with associated irqs, so this version should work with virtio-block, virtio-net, virtio-console, etc.

At the moment i've only figured out the correct qemu command line for virtio-block, but virtio-net and virtio-console should work assuming the correct qemu command line.

This is the command line I used to attach a virtio-block device:

`qemu-system-riscv64 -nographic -machine virt -kernel bbl-1.10-virtio -drive file=root.bin,format=raw,id=hd0 -device virtio-blk-device,drive=hd0`

You can see the device-tree here:

```
**** fdtdump is a low-level debugging tool, not meant for general use.
**** If you want to decompile a dtb, you probably want
****     dtc -I dtb -O dts <filename>

/dts-v1/;
// magic:               0xd00dfeed
// totalsize:           0x10000 (65536)
// off_dt_struct:       0x40
// off_dt_strings:      0x7b8
// off_mem_rsvmap:      0x30
// version:             17
// last_comp_version:   16
// boot_cpuid_phys:     0x0
// size_dt_strings:     0x114
// size_dt_struct:      0x778

/ {
    #address-cells = <0x00000002>;
    #size-cells = <0x00000002>;
    compatible = "riscv-virtio";
    model = "riscv-virtio,qemu";
    virtio_mmio@10007000 {
        interrupts = <0x00000008>;
        interrupt-parent = <0x00000002>;
        reg = <0x00000000 0x10007000 0x00000000 0x00001000>;
        compatible = "virtio,mmio";
    };
    virtio_mmio@10006000 {
        interrupts = <0x00000007>;
        interrupt-parent = <0x00000002>;
        reg = <0x00000000 0x10006000 0x00000000 0x00001000>;
        compatible = "virtio,mmio";
    };
    virtio_mmio@10005000 {
        interrupts = <0x00000006>;
        interrupt-parent = <0x00000002>;
        reg = <0x00000000 0x10005000 0x00000000 0x00001000>;
        compatible = "virtio,mmio";
    };
    virtio_mmio@10004000 {
        interrupts = <0x00000005>;
        interrupt-parent = <0x00000002>;
        reg = <0x00000000 0x10004000 0x00000000 0x00001000>;
        compatible = "virtio,mmio";
    };
    virtio_mmio@10003000 {
        interrupts = <0x00000004>;
        interrupt-parent = <0x00000002>;
        reg = <0x00000000 0x10003000 0x00000000 0x00001000>;
        compatible = "virtio,mmio";
    };
    virtio_mmio@10002000 {
        interrupts = <0x00000003>;
        interrupt-parent = <0x00000002>;
        reg = <0x00000000 0x10002000 0x00000000 0x00001000>;
        compatible = "virtio,mmio";
    };
    virtio_mmio@10001000 {
        interrupts = <0x00000002>;
        interrupt-parent = <0x00000002>;
        reg = <0x00000000 0x10001000 0x00000000 0x00001000>;
        compatible = "virtio,mmio";
    };
    virtio_mmio@10000000 {
        interrupts = <0x00000001>;
        interrupt-parent = <0x00000002>;
        reg = <0x00000000 0x10000000 0x00000000 0x00001000>;
        compatible = "virtio,mmio";
    };
    cpus {
        #address-cells = <0x00000001>;
        #size-cells = <0x00000000>;
        timebase-frequency = <0x00989680>;
        cpu@0 {
            device_type = "cpu";
            reg = <0x00000000>;
            status = "okay";
            compatible = "riscv";
            riscv,isa = "rv64imafdc";
            mmu-type = "riscv,sv48";
            clock-frequency = <0x3b9aca00>;
            interrupt-controller {
                #interrupt-cells = <0x00000001>;
                interrupt-controller;
                compatible = "riscv,cpu-intc";
                linux,phandle = <0x00000001>;
                phandle = <0x00000001>;
            };
        };
    };
    memory@80000000 {
        device_type = "memory";
        reg = <0x00000000 0x80000000 0x00000000 0x08000000>;
    };
    soc {
        #address-cells = <0x00000002>;
        #size-cells = <0x00000002>;
        compatible = "riscv-virtio-soc";
        ranges;
        interrupt-controller@c000000 {
            linux,phandle = <0x00000002>;
            phandle = <0x00000002>;
            riscv,ndev = <0x00000008>;
            riscv,max-priority = <0x00000007>;
            reg-names = "control";
            reg = <0x00000000 0x0c000000 0x00000000 0x04000000>;
            interrupts-extended = <0x00000001 0x0000000b 0x00000001 0x00000009>;
            interrupt-controller;
            compatible = "riscv,plic0";
            #interrupt-cells = <0x00000001>;
        };
        clint@2000000 {
            interrupts-extended = <0x00000001 0x00000003 0x00000001 0x00000007>;
            reg = <0x00000000 0x02000000 0x00000000 0x00010000>;
            compatible = "riscv,clint0";
        };
    };
    htif {
        compatible = "ucb,htif0";
    };
};
```

I'm still debugging an issue with the new PLIC however it works well enough to boot from virtio-block. If you figure out the correct qemu command line, you should be able to use virtio-net and virtio-console.

```
$ qemu-system-riscv64 -nographic -machine virt -kernel bbl-1.10-virtio -drive file=root.bin,format=raw,id=hd0 -device virtio-blk-device,drive=hd0
qemu-system-riscv64: plic: invalid register write: 00002088
qemu-system-riscv64: plic: invalid register write: 0000208c
[    0.000000] OF: fdt: Ignoring memory range 0x80000000 - 0x80200000
[    0.000000] Linux version 4.14.0-00029-g1c6baf1066cc-dirty (mclark@minty) (gcc version 7.1.1 20170509 (GCC)) #27 Sat Dec 9 19:53:20 NZDT 2017
[    0.000000] bootconsole [early0] enabled
[    0.000000] Zone ranges:
[    0.000000]   DMA      [mem 0x0000000080200000-0x0000000087ffffff]
[    0.000000]   Normal   empty
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000080200000-0x0000000087ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000080200000-0x0000000087ffffff]
[    0.000000] elf_hwcap is 0x112d
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 31815
[    0.000000] Kernel command line: root=/dev/vda ro 
[    0.000000] PID hash table entries: 512 (order: 0, 4096 bytes)
[    0.000000] Dentry cache hash table entries: 16384 (order: 5, 131072 bytes)
[    0.000000] Inode-cache hash table entries: 8192 (order: 4, 65536 bytes)
[    0.000000] Sorting __ex_table...
[    0.000000] Memory: 123260K/129024K available (2069K kernel code, 171K rwdata, 560K rodata, 84K init, 761K bss, 5764K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=64, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 0, nr_irqs: 0, preallocated irqs: 0
[    0.000000] riscv,cpu_intc,0: 64 local interrupts mapped
[    0.000000] riscv,plic0,c000000: mapped 8 interrupts to 1/2 handlers
[    0.000000] clocksource: riscv_clocksource: mask: 0xffffffffffffffff max_cycles: 0x24e6a1710, max_idle_ns: 440795202120 ns
[    0.000000] console [hvc0] enabled
[    0.000000] console [hvc0] enabled
[    0.000000] bootconsole [early0] disabled
[    0.000000] bootconsole [early0] disabled
[    0.000000] Calibrating delay loop (skipped), value calculated using timer frequency.. 20.00 BogoMIPS (lpj=100000)
[    0.000000] pid_max: default: 32768 minimum: 301
[    0.000000] Mount-cache hash table entries: 512 (order: 0, 4096 bytes)
[    0.000000] Mountpoint-cache hash table entries: 512 (order: 0, 4096 bytes)
[    0.040000] devtmpfs: initialized
[    0.050000] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.050000] futex hash table entries: 256 (order: 0, 6144 bytes)
[    0.050000] random: get_random_u32 called from bucket_table_alloc+0x80/0x1f4 with crng_init=0
[    0.060000] NET: Registered protocol family 16
[    0.070000] vgaarb: loaded
[    0.070000] clocksource: Switched to clocksource riscv_clocksource
[    0.080000] NET: Registered protocol family 2
[    0.080000] TCP established hash table entries: 1024 (order: 1, 8192 bytes)
[    0.080000] TCP bind hash table entries: 1024 (order: 1, 8192 bytes)
[    0.080000] TCP: Hash tables configured (established 1024 bind 1024)
[    0.080000] UDP hash table entries: 256 (order: 1, 8192 bytes)
[    0.080000] UDP-Lite hash table entries: 256 (order: 1, 8192 bytes)
[    0.080000] NET: Registered protocol family 1
[    0.090000] workingset: timestamp_bits=62 max_order=15 bucket_order=0
[    0.120000] 9p: Installing v9fs 9p2000 file system support
[    0.120000] random: fast init done
[    0.140000] io scheduler noop registered
[    0.140000] io scheduler cfq registered (default)
[    0.140000] io scheduler mq-deadline registered
[    0.140000] io scheduler kyber registered
[    0.310000] random: crng init done
[    0.310000] irq 8: nobody cared (try booting with the "irqpoll" option)
[    0.310000] CPU: 0 PID: 4 Comm: kworker/0:0H Not tainted 4.14.0-00029-g1c6baf1066cc-dirty #27
[    0.310000] Workqueue: kblockd blk_mq_run_work_fn
[    0.310000] Call Trace:
[    0.310000] [<ffffffe000016a30>] walk_stackframe+0x0/0x7a
[    0.310000] [<ffffffe00003b282>] __report_bad_irq.isra.0+0x2c/0xb8
[    0.310000] [<ffffffe00003b438>] note_interrupt+0xe2/0x226
[    0.310000] [<ffffffe000039acc>] handle_irq_event_percpu+0x34/0x44
[    0.310000] [<ffffffe0001763da>] virtio_queue_rq+0x1f8/0x20e
[    0.310000] [<ffffffe000039afc>] handle_irq_event+0x20/0x3c
[    0.310000] [<ffffffe00003bc90>] handle_simple_irq+0x4a/0x72
[    0.310000] [<ffffffe000039470>] generic_handle_irq+0x1c/0x2c
[    0.310000] [<ffffffe00013cce4>] plic_chained_handle_irq+0x46/0xee
[    0.310000] [<ffffffe000122c78>] blk_mq_dispatch_rq_list+0x264/0x2ba
[    0.310000] [<ffffffe000039470>] generic_handle_irq+0x1c/0x2c
[    0.310000] [<ffffffe00013d3a2>] riscv_intc_irq+0x50/0x8c
[    0.310000] [<ffffffe000015c30>] ret_from_syscall+0xa/0xe
[    0.310000] [<ffffffe000122c78>] blk_mq_dispatch_rq_list+0x264/0x2ba
[    0.310000] [<ffffffe000122c0e>] blk_mq_dispatch_rq_list+0x1fa/0x2ba
[    0.310000] [<ffffffe00003ee0e>] __internal_add_timer.constprop.9+0x20/0x2c
[    0.310000] [<ffffffe000121e2e>] __blk_mq_delay_run_hw_queue+0xdc/0xe0
[    0.310000] [<ffffffe000122c0e>] blk_mq_dispatch_rq_list+0x1fa/0x2ba
[    0.310000] [<ffffffe000125b9a>] blk_mq_sched_dispatch_requests+0xae/0x136
[    0.310000] [<ffffffe000025f26>] process_one_work+0x156/0x2c0
[    0.310000] [<ffffffe0000260f4>] worker_thread+0x64/0x3d8
[    0.310000] [<ffffffe00002608c>] process_one_work+0x2bc/0x2c0
[    0.310000] [<ffffffe00002a3fa>] kthread+0xe8/0xfe
[    0.310000] [<ffffffe00002a30e>] __kthread_create_on_node+0xe0/0xe4
[    0.310000] [<ffffffe000015c30>] ret_from_syscall+0xa/0xe
[    0.310000] handlers:
[    0.310000] [<ffffffe00015ade8>] vm_interrupt
[    0.310000] Disabling IRQ #8
[    1.040000] 9pnet: Installing 9P2000 support
[    1.200000] EXT4-fs (vda): couldn't mount as ext3 due to feature incompatibilities
[    1.310000] EXT4-fs (vda): mounting ext2 file system using the ext4 subsystem
[    1.530000] EXT4-fs (vda): mounted filesystem without journal. Opts: (null)
[    1.530000] VFS: Mounted root (ext2 filesystem) readonly on device 254:0.
[    1.750000] devtmpfs: mounted
[    1.760000] Freeing unused kernel memory: 84K
[    1.760000] This architecture does not have kernel memory protection.

Welcome to RISCVEMU (64 bits)
sh-4.3# cat /proc/interrupts 
           CPU0       
  8:     100000  riscv,plic0,c000000   8  virtio0
sh-4.3# cat /proc/partitions 
major minor  #blocks  name

 254        0     262144 vda
sh-4.3# 
```

michaeljclark commented:
For some reason, the virtio device IRQ is expected to be disabled by default, and the virtio device is sending interrupts, and the kernel complains, which leads to a harmless backtrace. The kernel later re-enables the IRQ and proceeds as normal. I haven't figured out if this is an issue with the QEMU PLIC implementation or the linux-kernel PLIC irq chip driver. It may be the QEMU PLIC priority handling code.

Its possible to uncomment `RISCV_DEBUG_PLIC` in `hw/riscv/sifive_plic.c` to look at the chatter between the kernel and the qemu emulated PLIC.

In any case it is not a regression as there is no PLIC in master, and it works well enough to boot from a virtio block device. There are probably still a few bugs in the PLIC, but they are not regressions.

If you'd like to test virtio you can pull this branch:

- https://github.com/michaeljclark/riscv-qemu/tree/plic

@sagark  I think it is safe to merge but you could wait a couple of days to see if I can shake out any remaining bugs in the PLIC implementation... The format string build failure with -Werror has been fixed...

michaeljclark commented:
The PLIC on this board only has 63 IRQs so these messages are valid:

```
qemu-system-riscv64: plic: invalid register write: 00002088
qemu-system-riscv64: plic: invalid register write: 0000208c
```

0x0c002080 and 0x0c002084 cover the enables for 63 IRQS.

`linux-kernel/drivers/irqchip/irq-riscv-plic.c` is hard coded for 127 IRQs

```
#define ENABLE_PER_HART		0x80
```

So I'll update the PLICs to have 127 sources.

michaeljclark commented:
Okay, that last patch fixed the warning messages from the PLIC by making the PLIC support 127 IRQs. Ideally spike should read the ndevs property from device-tree.

The only remaining issue is to figure out why the IRQ gets enabled early and results in the _apparently harmless_ backtrace. I'll enable `RISCV_PLIC_DEBUG` and see what's going on.

```
$ qemu-system-riscv64 -nographic -machine virt -kernel bbl-1.10-virtio -drive file=root.bin,format=raw,id=hd0 -device virtio-blk-device,drive=hd0
[    0.000000] OF: fdt: Ignoring memory range 0x80000000 - 0x80200000
[    0.000000] Linux version 4.14.0-00029-g1c6baf1066cc-dirty (mclark@minty) (gcc version 7.1.1 20170509 (GCC)) #27 Sat Dec 9 19:53:20 NZDT 2017
[    0.000000] bootconsole [early0] enabled
[    0.000000] Zone ranges:
[    0.000000]   DMA      [mem 0x0000000080200000-0x0000000087ffffff]
[    0.000000]   Normal   empty
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000080200000-0x0000000087ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000080200000-0x0000000087ffffff]
[    0.000000] elf_hwcap is 0x112d
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 31815
[    0.000000] Kernel command line: root=/dev/vda ro 
[    0.000000] PID hash table entries: 512 (order: 0, 4096 bytes)
[    0.000000] Dentry cache hash table entries: 16384 (order: 5, 131072 bytes)
[    0.000000] Inode-cache hash table entries: 8192 (order: 4, 65536 bytes)
[    0.000000] Sorting __ex_table...
[    0.000000] Memory: 123260K/129024K available (2069K kernel code, 171K rwdata, 560K rodata, 84K init, 761K bss, 5764K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=64, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 0, nr_irqs: 0, preallocated irqs: 0
[    0.000000] riscv,cpu_intc,0: 64 local interrupts mapped
[    0.000000] riscv,plic0,c000000: mapped 8 interrupts to 1/2 handlers
[    0.000000] clocksource: riscv_clocksource: mask: 0xffffffffffffffff max_cycles: 0x24e6a1710, max_idle_ns: 440795202120 ns
[    0.000000] console [hvc0] enabled
[    0.000000] console [hvc0] enabled
[    0.000000] bootconsole [early0] disabled
[    0.000000] bootconsole [early0] disabled
[    0.000000] Calibrating delay loop (skipped), value calculated using timer frequency.. 20.00 BogoMIPS (lpj=100000)
[    0.000000] pid_max: default: 32768 minimum: 301
[    0.000000] Mount-cache hash table entries: 512 (order: 0, 4096 bytes)
[    0.000000] Mountpoint-cache hash table entries: 512 (order: 0, 4096 bytes)
[    0.040000] devtmpfs: initialized
[    0.050000] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.050000] futex hash table entries: 256 (order: 0, 6144 bytes)
[    0.050000] random: get_random_u32 called from bucket_table_alloc+0x80/0x1f4 with crng_init=0
[    0.050000] NET: Registered protocol family 16
[    0.070000] vgaarb: loaded
[    0.080000] clocksource: Switched to clocksource riscv_clocksource
[    0.080000] NET: Registered protocol family 2
[    0.090000] TCP established hash table entries: 1024 (order: 1, 8192 bytes)
[    0.090000] TCP bind hash table entries: 1024 (order: 1, 8192 bytes)
[    0.090000] TCP: Hash tables configured (established 1024 bind 1024)
[    0.090000] UDP hash table entries: 256 (order: 1, 8192 bytes)
[    0.090000] UDP-Lite hash table entries: 256 (order: 1, 8192 bytes)
[    0.090000] NET: Registered protocol family 1
[    0.100000] workingset: timestamp_bits=62 max_order=15 bucket_order=0
[    0.130000] 9p: Installing v9fs 9p2000 file system support
[    0.140000] random: fast init done
[    0.150000] io scheduler noop registered
[    0.150000] io scheduler cfq registered (default)
[    0.150000] io scheduler mq-deadline registered
[    0.150000] io scheduler kyber registered
[    0.320000] random: crng init done
[    0.320000] irq 8: nobody cared (try booting with the "irqpoll" option)
[    0.320000] CPU: 0 PID: 4 Comm: kworker/0:0H Not tainted 4.14.0-00029-g1c6baf1066cc-dirty #27
[    0.320000] Workqueue: kblockd blk_mq_run_work_fn
[    0.320000] Call Trace:
[    0.320000] [<ffffffe000016a30>] walk_stackframe+0x0/0x7a
[    0.320000] [<ffffffe00003b282>] __report_bad_irq.isra.0+0x2c/0xb8
[    0.320000] [<ffffffe00003b438>] note_interrupt+0xe2/0x226
[    0.320000] [<ffffffe000039acc>] handle_irq_event_percpu+0x34/0x44
[    0.320000] [<ffffffe000039afc>] handle_irq_event+0x20/0x3c
[    0.320000] [<ffffffe00003bc90>] handle_simple_irq+0x4a/0x72
[    0.320000] [<ffffffe000039470>] generic_handle_irq+0x1c/0x2c
[    0.320000] [<ffffffe00013cce4>] plic_chained_handle_irq+0x46/0xee
[    0.320000] [<ffffffe000122c82>] blk_mq_dispatch_rq_list+0x26e/0x2ba
[    0.320000] [<ffffffe000039470>] generic_handle_irq+0x1c/0x2c
[    0.320000] [<ffffffe00013d3a2>] riscv_intc_irq+0x50/0x8c
[    0.320000] [<ffffffe000015c30>] ret_from_syscall+0xa/0xe
[    0.320000] [<ffffffe000122c82>] blk_mq_dispatch_rq_list+0x26e/0x2ba
[    0.320000] [<ffffffe000122c0e>] blk_mq_dispatch_rq_list+0x1fa/0x2ba
[    0.320000] [<ffffffe00003ee0e>] __internal_add_timer.constprop.9+0x20/0x2c
[    0.320000] [<ffffffe000121e2e>] __blk_mq_delay_run_hw_queue+0xdc/0xe0
[    0.320000] [<ffffffe000122c0e>] blk_mq_dispatch_rq_list+0x1fa/0x2ba
[    0.320000] [<ffffffe000125b9a>] blk_mq_sched_dispatch_requests+0xae/0x136
[    0.320000] [<ffffffe000025f26>] process_one_work+0x156/0x2c0
[    0.320000] [<ffffffe0000260f4>] worker_thread+0x64/0x3d8
[    0.320000] [<ffffffe00002608c>] process_one_work+0x2bc/0x2c0
[    0.320000] [<ffffffe00002a3fa>] kthread+0xe8/0xfe
[    0.320000] [<ffffffe00002a30e>] __kthread_create_on_node+0xe0/0xe4
[    0.320000] [<ffffffe000015c30>] ret_from_syscall+0xa/0xe
[    0.320000] handlers:
[    0.320000] [<ffffffe00015ade8>] vm_interrupt
[    0.320000] Disabling IRQ #8
[    1.060000] 9pnet: Installing 9P2000 support
[    1.200000] EXT4-fs (vda): couldn't mount as ext3 due to feature incompatibilities
[    1.310000] EXT4-fs (vda): mounting ext2 file system using the ext4 subsystem
[    1.530000] EXT4-fs (vda): mounted filesystem without journal. Opts: (null)
[    1.530000] VFS: Mounted root (ext2 filesystem) readonly on device 254:0.
[    1.750000] devtmpfs: mounted
[    1.760000] Freeing unused kernel memory: 84K
[    1.760000] This architecture does not have kernel memory protection.

Welcome to RISCVEMU (64 bits)
sh-4.3# 
```

michaeljclark commented:
When enabling PLIC debugging in QEMU it seems the IRQ is enabled when the kernel complains. You can see here that IRQ 1 through 7 are disabled and IRQ 8 is enabled. You can even see the plic write that enables IRQ 8.

```
sifive_plic_raise_irq: irq=8
plic: write priority: irq=8 priority=1
plic: read enable: hart0-S word=0 value=fffffe01
plic: write enable: hart0-S word=0 value=ffffff01
sifive_plic_raise_irq: irq=8
sifive_plic_raise_irq: irq=8 -> hart0-S
plic: read claim: hart0-S irq=8
pending       : 00000000000000000000000000000000
claimed       : 00000000000000000000000000000100
hart0-M enable: 00000000000000000000000000000000
hart0-S enable: ffffffffffffffffffffffffffffff01
sifive_plic_raise_irq: irq=8
sifive_plic_lower_irq: irq=8
sifive_plic_lower_irq: irq=8 -> hart0-M
plic: write claim: hart0-S irq=8
```

there are several interrupts and then the kernel disables the IRQ. virtio must then be reverting to polling the MMIO aperture as interrupts are no longer delivered after the IRQ has been disabled, but the virtio block device continues to work. You can see the kernel set the priority to 0 and set the enable bit to 0 for IRQ 8.

```
[    0.330000] irq 8: nobody cared (try booting with the "irqpoll" option)
[    0.330000] CPU: 0 PID: 4 Comm: kworker/0:0H Not tainted 4.14.0-00029-g1c6baf1066cc-dirty #27
[    0.330000] Workqueue: kblockd blk_mq_run_work_fn
[    0.330000] Call Trace:
[    0.330000] [<ffffffe000016a30>] walk_stackframe+0x0/0x7a
[    0.330000] [<ffffffe00003b282>] __report_bad_irq.isra.0+0x2c/0xb8
[    0.330000] [<ffffffe00003b438>] note_interrupt+0xe2/0x226
[    0.330000] [<ffffffe000039acc>] handle_irq_event_percpu+0x34/0x44
[    0.330000] [<ffffffe000039afc>] handle_irq_event+0x20/0x3c
[    0.330000] [<ffffffe00003bc90>] handle_simple_irq+0x4a/0x72
[    0.330000] [<ffffffe000039470>] generic_handle_irq+0x1c/0x2c
[    0.330000] [<ffffffe00013cce4>] plic_chained_handle_irq+0x46/0xee
[    0.330000] [<ffffffe000039470>] generic_handle_irq+0x1c/0x2c
[    0.330000] [<ffffffe00013d3a2>] riscv_intc_irq+0x50/0x8c
[    0.330000] [<ffffffe000015c30>] ret_from_syscall+0xa/0xe
[    0.330000] [<ffffffe00013d3aa>] riscv_intc_irq+0x58/0x8c
[    0.330000] [<ffffffe000121e2e>] __blk_mq_delay_run_hw_queue+0xdc/0xe0
[    0.330000] [<ffffffe000121e2e>] __blk_mq_delay_run_hw_queue+0xdc/0xe0
[    0.330000] [<ffffffe000122c78>] blk_mq_dispatch_rq_list+0x264/0x2ba
[    0.330000] [<ffffffe00013d3aa>] riscv_intc_irq+0x58/0x8c
[    0.330000] [<ffffffe000015c30>] ret_from_syscall+0xa/0xe
[    0.330000] [<ffffffe000122c78>] blk_mq_dispatch_rq_list+0x264/0x2ba
[    0.330000] [<ffffffe000122c0e>] blk_mq_dispatch_rq_list+0x1fa/0x2ba
[    0.330000] [<ffffffe00003ee0e>] __internal_add_timer.constprop.9+0x20/0x2c
[    0.330000] [<ffffffe000121e2e>] __blk_mq_delay_run_hw_queue+0xdc/0xe0
[    0.330000] [<ffffffe000122c0e>] blk_mq_dispatch_rq_list+0x1fa/0x2ba
[    0.330000] [<ffffffe000125b9a>] blk_mq_sched_dispatch_requests+0xae/0x136
[    0.330000] [<ffffffe000025f26>] process_one_work+0x156/0x2c0
[    0.330000] [<ffffffe0000260f4>] worker_thread+0x64/0x3d8
[    0.330000] [<ffffffe00002608c>] process_one_work+0x2bc/0x2c0
[    0.330000] [<ffffffe00002a3fa>] kthread+0xe8/0xfe
[    0.330000] [<ffffffe00002a30e>] __kthread_create_on_node+0xe0/0xe4
[    0.330000] [<ffffffe000015c30>] ret_from_syscall+0xa/0xe
[    0.330000] handlers:
[    0.330000] [<ffffffe00015ade8>] vm_interrupt
[    0.330000] Disabling IRQ #8
plic: write priority: irq=8 priority=0
plic: read enable: hart0-S word=0 value=ffffff01
plic: write enable: hart0-S word=0 value=fffffe01
```

michaeljclark commented:
I was able to get networking working with this command:

```
sudo qemu-system-riscv64 -nographic \
    -machine virt -kernel bbl-1.10-virtio \
    -drive file=root.bin,format=raw,id=hd0 \
    -device virtio-blk-device,drive=hd0 \
    -netdev type=tap,script=./ifup,downscript=./ifdown,id=net0 \
    -device virtio-net-device,netdev=net0
```

This configuration was done on the host (installed tuntap using brew):

```
# create bridge
sudo ifconfig bridge1 create
sudo ifconfig bridge1 192.168.100.1/24

# create ifup script
cat > ifup <<EOF
#!/bin/bash

ifconfig bridge1 addm $1
EOF
chmod +x ifup

# create ifdown script
cat > ifdown <<EOF
#!/bin/bash

ifconfig bridge1 deletem $1
EOF
chmod +x idown
```

in linux:

```
sh-4.3# cat /proc/interrupts 
           CPU0       
  7:     100000  riscv,plic0,c000000   7  virtio1
  8:     100000  riscv,plic0,c000000   8  virtio0
sh-4.3# ifconfig eth0 192.168.100.2
sh-4.3# ping 192.168.100.1
PING 192.168.100.1 (192.168.100.1): 56 data bytes
64 bytes from 192.168.100.1: seq=0 ttl=64 time=101.885 ms
64 bytes from 192.168.100.1: seq=1 ttl=64 time=79.469 ms
64 bytes from 192.168.100.1: seq=2 ttl=64 time=60.118 ms
^C
--- 192.168.100.1 ping statistics ---
3 packets transmitted, 3 packets received, 0% packet loss
round-trip min/avg/max = 60.118/80.490/101.885 ms
sh-4.3# uname -a
Linux ucbvax 4.14.0-00029-g1c6baf1066cc-dirty #30 Sun Dec 10 14:21:02 NZDT 2017 riscv64 GNU/Linux
sh-4.3# 
```

davidlt commented:
Regarding virtconsole, I found this documents:
- http://nairobi-embedded.org/qemu_virtio.html#virtio-serial
- http://nairobi-embedded.org/qemu_character_devices.html#guest-io-via-virtio-serial

Could you give a try?

michaeljclark commented:
@palmer-dabbelt i've been debugging the software PLIC emulation and as far as I can tell, the virtual PLIC is sending interrupts to the kernel correctly, when they have been unmasked.

The backtrace seems to indicate a problem with irq domains. The PLIC is forwarding the assigned IRQ to generic_handle_irq. It could be an issue with the device-tree. I might take a look at the arm virt board device tree to see whether we are missing something. I guess irq domains are working in this kernel revision. It's puzzling.

It is clear from the debug that the IRQ has been enabled by the kernel. It seems to reach a threshold of 100000 interrupts before being disabled by the kernel. VirtIO continues to work, so the kernel driver must be polling the MMIO aperture on each clock tick. e.g. CONFIG_HZ=100

```
sh-4.3# cat /proc/interrupts 
           CPU0       
  8:     100000  riscv,plic0,c000000   8  virtio0
```

Now, we have two IRQ domains here, the PLIC and the local IRQ domain. Shouldn't the local IRQ be 9 "Supervisor external interrupt" ?

michaeljclark commented:
In any case, I think this PR is clean enough to merge despite the IRQ domain issue.

We can merge now or wait until we figure out the irq domain issue. The VirtIO works via polling, and the PLIC seems to be working correctly based on debug logs. i.e. it fixes some issues in master.

@palmer-dabbelt shall we wait until we've figured out why the IRQ supression kicks in?

Note: when enabling RISCV_DEBUG_PLIC. This backtrace happens immediately after the kernel sets a non zero priority and the enable bit for hart0-S mode context IRQ 8.

```
[    0.330000] irq 8: nobody cared (try booting with the "irqpoll" option)
[    0.330000] CPU: 0 PID: 4 Comm: kworker/0:0H Not tainted 4.14.0-00029-g1c6baf1066cc-dirty #30
[    0.330000] Workqueue: kblockd blk_mq_run_work_fn
[    0.330000] Call Trace:
[    0.330000] [<ffffffe000016a30>] walk_stackframe+0x0/0x7a
[    0.330000] [<ffffffe00003b282>] __report_bad_irq.isra.0+0x2c/0xb8
[    0.330000] [<ffffffe00003b438>] note_interrupt+0xe2/0x226
[    0.330000] [<ffffffe000039acc>] handle_irq_event_percpu+0x34/0x44
[    0.330000] [<ffffffe000039afc>] handle_irq_event+0x20/0x3c
[    0.330000] [<ffffffe00003bc90>] handle_simple_irq+0x4a/0x72
[    0.330000] [<ffffffe000039470>] generic_handle_irq+0x1c/0x2c
[    0.330000] [<ffffffe00013cce4>] plic_chained_handle_irq+0x46/0xee
[    0.330000] [<ffffffe000039470>] generic_handle_irq+0x1c/0x2c
[    0.330000] [<ffffffe00013d3a2>] riscv_intc_irq+0x50/0x8c
[    0.330000] [<ffffffe000015c30>] ret_from_syscall+0xa/0xe
[    0.330000] [<ffffffe00013d3aa>] riscv_intc_irq+0x58/0x8c
[    0.330000] [<ffffffe000121e2e>] __blk_mq_delay_run_hw_queue+0xdc/0xe0
[    0.330000] [<ffffffe000121e2e>] __blk_mq_delay_run_hw_queue+0xdc/0xe0
[    0.330000] [<ffffffe000122b64>] blk_mq_dispatch_rq_list+0x150/0x2ba
[    0.330000] [<ffffffe00013d3aa>] riscv_intc_irq+0x58/0x8c
[    0.330000] [<ffffffe000015c30>] ret_from_syscall+0xa/0xe
[    0.330000] [<ffffffe000122b64>] blk_mq_dispatch_rq_list+0x150/0x2ba
[    0.330000] [<ffffffe000122c0e>] blk_mq_dispatch_rq_list+0x1fa/0x2ba
[    0.330000] [<ffffffe00003ee0e>] __internal_add_timer.constprop.9+0x20/0x2c
[    0.330000] [<ffffffe000121e2e>] __blk_mq_delay_run_hw_queue+0xdc/0xe0
[    0.330000] [<ffffffe000122c0e>] blk_mq_dispatch_rq_list+0x1fa/0x2ba
[    0.330000] [<ffffffe000125b9a>] blk_mq_sched_dispatch_requests+0xae/0x136
[    0.330000] [<ffffffe000025f26>] process_one_work+0x156/0x2c0
[    0.330000] [<ffffffe0000260f4>] worker_thread+0x64/0x3d8
[    0.330000] [<ffffffe00002608c>] process_one_work+0x2bc/0x2c0
[    0.330000] [<ffffffe00002a3fa>] kthread+0xe8/0xfe
[    0.330000] [<ffffffe00002a30e>] __kthread_create_on_node+0xe0/0xe4
[    0.330000] [<ffffffe000015c30>] ret_from_syscall+0xa/0xe
[    0.330000] handlers:
[    0.330000] [<ffffffe00015ade8>] vm_interrupt
[    0.330000] Disabling IRQ #8
```

palmer-dabbelt commented:
Let me give it a shot.  The tip of this branch manifests the problem?

michaeljclark commented:
Yap

palmer-dabbelt commented:
@michaeljclark I've got a circular pull request going on here...
https://github.com/michaeljclark/riscv-qemu/pull/1

michaeljclark commented:
Palmer solved the interrupt problem. We were not using the level parameter correctly in the PLIC irq request callback, and that was causing spurious interrupts to be delivered to Linux. Then we solved a second issue with cpu_has_work/WFI that was causing dropped interrupts. This PR seems good now.

@palmer-dabbelt this seems stable to me now. shall we merge it?

I think we should. I no longer get dropped console characters and VirtIO seems stable.

```
$ qemu-system-riscv64 -nographic -machine virt -kernel bbl-1.10-virtio -drive file=root.bin,format=raw,id=hd0 -device virtio-blk-device,drive=hd0 -netdev type=user,id=net0,hostfwd=tcp::5555-:22 -device virtio-net-device,netdev=net0
[    0.000000] OF: fdt: Ignoring memory range 0x80000000 - 0x80200000
[    0.000000] Linux version 4.14.0-00029-g1c6baf1066cc (mclark@minty) (gcc version 7.1.1 20170509 (GCC)) #34 Tue Dec 12 14:43:28 NZDT 2017
[    0.000000] bootconsole [early0] enabled
[    0.000000] Zone ranges:
[    0.000000]   DMA      [mem 0x0000000080200000-0x0000000087ffffff]
[    0.000000]   Normal   empty
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000080200000-0x0000000087ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000080200000-0x0000000087ffffff]
[    0.000000] elf_hwcap is 0x112d
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 31815
[    0.000000] Kernel command line: root=/dev/vda ro 
[    0.000000] PID hash table entries: 512 (order: 0, 4096 bytes)
[    0.000000] Dentry cache hash table entries: 16384 (order: 5, 131072 bytes)
[    0.000000] Inode-cache hash table entries: 8192 (order: 4, 65536 bytes)
[    0.000000] Sorting __ex_table...
[    0.000000] Memory: 123260K/129024K available (2069K kernel code, 171K rwdata, 560K rodata, 84K init, 761K bss, 5764K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=64, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 0, nr_irqs: 0, preallocated irqs: 0
[    0.000000] riscv,cpu_intc,0: 64 local interrupts mapped
[    0.000000] riscv,plic0,c000000: mapped 8 interrupts to 1/2 handlers
[    0.000000] clocksource: riscv_clocksource: mask: 0xffffffffffffffff max_cycles: 0x24e6a1710, max_idle_ns: 440795202120 ns
[    0.000000] console [hvc0] enabled
[    0.000000] console [hvc0] enabled
[    0.000000] bootconsole [early0] disabled
[    0.000000] bootconsole [early0] disabled
[    0.000000] Calibrating delay loop (skipped), value calculated using timer frequency.. 20.00 BogoMIPS (lpj=100000)
[    0.000000] pid_max: default: 32768 minimum: 301
[    0.000000] Mount-cache hash table entries: 512 (order: 0, 4096 bytes)
[    0.010000] Mountpoint-cache hash table entries: 512 (order: 0, 4096 bytes)
[    0.050000] devtmpfs: initialized
[    0.050000] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.060000] futex hash table entries: 256 (order: 0, 6144 bytes)
[    0.060000] random: get_random_u32 called from bucket_table_alloc+0x80/0x1f4 with crng_init=0
[    0.060000] NET: Registered protocol family 16
[    0.080000] vgaarb: loaded
[    0.080000] clocksource: Switched to clocksource riscv_clocksource
[    0.090000] NET: Registered protocol family 2
[    0.090000] TCP established hash table entries: 1024 (order: 1, 8192 bytes)
[    0.090000] TCP bind hash table entries: 1024 (order: 1, 8192 bytes)
[    0.090000] TCP: Hash tables configured (established 1024 bind 1024)
[    0.090000] UDP hash table entries: 256 (order: 1, 8192 bytes)
[    0.090000] UDP-Lite hash table entries: 256 (order: 1, 8192 bytes)
[    0.100000] NET: Registered protocol family 1
[    0.100000] workingset: timestamp_bits=62 max_order=15 bucket_order=0
[    0.130000] 9p: Installing v9fs 9p2000 file system support
[    0.140000] random: fast init done
[    0.150000] io scheduler noop registered
[    0.150000] io scheduler cfq registered (default)
[    0.150000] io scheduler mq-deadline registered
[    0.160000] io scheduler kyber registered
[    0.340000] 9pnet: Installing 9P2000 support
[    0.350000] EXT4-fs (vda): couldn't mount as ext3 due to feature incompatibilities
[    0.360000] EXT4-fs (vda): mounting ext2 file system using the ext4 subsystem
[    0.370000] EXT4-fs (vda): mounted filesystem without journal. Opts: (null)
[    0.370000] VFS: Mounted root (ext2 filesystem) readonly on device 254:0.
[    0.370000] devtmpfs: mounted
[    0.380000] Freeing unused kernel memory: 84K
[    0.380000] This architecture does not have kernel memory protection.
~ # cat /proc/interrupts 
           CPU0       
  7:          0  riscv,plic0,c000000   7  virtio1
  8:         72  riscv,plic0,c000000   8  virtio0
~ # ifconfig eth0 192.168.100.1
~ # ifconfig -a
eth0      Link encap:Ethernet  HWaddr 52:54:00:12:34:56  
          inet addr:192.168.100.1  Bcast:192.168.100.255  Mask:255.255.255.0
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

~ # cat /proc/partitions 
major minor  #blocks  name

 254        0       4096 vda
~ # 
```

michaeljclark commented:
This branch still uses SBI/HTIF console. We can add virtio console in another PR. It will be nice for folk to be able to use virtio-block and virtio-net.

virtio-console should work if we attach a device from the command line, just we may not be able to connect the console to stdout until we change the first chardev attached inside the machine.

Appreciate if anyone can test.

I'll work on putting kernel configs and command lines on the wiki tomorrow: https://github.com/riscv/riscv-qemu/wiki

```
$ more defconfig 
# CONFIG_COMPACTION is not set
CONFIG_HZ_100=y
CONFIG_PCI=y
CONFIG_CROSS_COMPILE="riscv64-unknown-elf-"
CONFIG_DEFAULT_HOSTNAME="ucbvax"
# CONFIG_CROSS_MEMORY_ATTACH is not set
# CONFIG_FHANDLE is not set
CONFIG_CGROUPS=y
CONFIG_NAMESPACES=y
CONFIG_EMBEDDED=y
# CONFIG_BLK_DEV_BSG is not set
CONFIG_PARTITION_ADVANCED=y
# CONFIG_EFI_PARTITION is not set
# CONFIG_IOSCHED_DEADLINE is not set
CONFIG_NET=y
CONFIG_UNIX=y
CONFIG_INET=y
# CONFIG_INET_XFRM_MODE_TRANSPORT is not set
# CONFIG_INET_XFRM_MODE_TUNNEL is not set
# CONFIG_INET_XFRM_MODE_BEET is not set
# CONFIG_INET_DIAG is not set
# CONFIG_IPV6 is not set
# CONFIG_WIRELESS is not set
CONFIG_NET_9P=y
CONFIG_NET_9P_VIRTIO=y
CONFIG_NET_9P_DEBUG=y
CONFIG_DEVTMPFS=y
CONFIG_DEVTMPFS_MOUNT=y
# CONFIG_FIRMWARE_IN_KERNEL is not set
CONFIG_VIRTIO_BLK=y
CONFIG_NETDEVICES=y
CONFIG_VIRTIO_NET=y
# CONFIG_ETHERNET is not set
# CONFIG_WLAN is not set
# CONFIG_INPUT_KEYBOARD is not set
# CONFIG_INPUT_MOUSE is not set
# CONFIG_VT is not set
CONFIG_DEVKMEM=y
CONFIG_VIRTIO_CONSOLE=y
# CONFIG_HW_RANDOM is not set
# CONFIG_HWMON is not set
# CONFIG_RC_CORE is not set
CONFIG_FB=y
# CONFIG_USB_SUPPORT is not set
CONFIG_VIRT_DRIVERS=y
CONFIG_VIRTIO_MMIO=y
# CONFIG_IOMMU_SUPPORT is not set
CONFIG_EXT4_FS=y
# CONFIG_FILE_LOCKING is not set
# CONFIG_DNOTIFY is not set
# CONFIG_PROC_PAGE_MONITOR is not set
CONFIG_TMPFS=y
# CONFIG_MISC_FILESYSTEMS is not set
CONFIG_9P_FS=y
CONFIG_CMDLINE_BOOL=y
CONFIG_CMDLINE="root=/dev/vda ro"
CONFIG_PRINTK_TIME=y
CONFIG_MESSAGE_LOGLEVEL_DEFAULT=3
CONFIG_DEBUG_SECTION_MISMATCH=y
# CONFIG_FRAME_POINTER is not set
# CONFIG_CRYPTO_HW is not set
```

davidlt commented:
Add some point we also need to update default riscv kernel configuration, e.g.
```
diff --git a/arch/riscv/configs/defconfig b/arch/riscv/configs/defconfig
index e69de29bb..e596ca4c6 100644
--- a/arch/riscv/configs/defconfig
+++ b/arch/riscv/configs/defconfig
@@ -0,0 +1,6 @@
+CONFIG_VIRTIO_BLK=y
+CONFIG_VIRTIO_NET=y
+CONFIG_VIRTIO_CONSOLE=y
+CONFIG_VIRTIO_PCI=y
+CONFIG_VIRTIO_BALLOON=y
+CONFIG_VIRTIO_MMIO=y
```

Also options you mentioned in the last commit.

michaeljclark commented:
@davidlt Agree. The development and testing part of VirtIO is pretty much complete.

Next is documentation. I need to spend some time documenting the various kernel config options and qemu command lines, etc on the wiki.

We also have to decide whether we will use HTIF or UART for console. UART seems to work better for keycode mapping. e.g. up arrow and down arrow work for history, but UART console depends on a patch to linux-kernel. I should make a PR to riscv-linux so that 8250/16550 compiles i.e.:

```
diff --git a/arch/riscv/include/asm/serial.h b/arch/riscv/include/asm/serial.h
new file mode 100644
index 000000000000..f51e0e10faa7
--- /dev/null
+++ b/arch/riscv/include/asm/serial.h
@@ -0,0 +1,10 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+#ifndef _ASM_SERIAL_H
+#define _ASM_SERIAL_H
+
+/*
+ * This assumes you have a 1.8432 MHz clock for your UART.
+ */
+#define BASE_BAUD (1843200 / 16)
+
+#endif /* _ASM_SERIAL_H */
```

davidlt commented:
@palmer-dabbelt could such patch land in later -rc?

I also quickly chatted with @palmer-dabbelt on IRC that our `arch/riscv/configs/defconfig` is empty. We should come up with some defaults everyone will expect (ext2, ext3, nfs, virtio stuff, some very well known drivers, etc). E.g., take arm64 one and remove all arm64 specific bits.

@palmer-dabbelt pointed that SDK is using this one: https://github.com/sifive/freedom-u-sdk/blob/master/conf/linux_defconfig

Thus we can also start from it.

michaeljclark commented:
We didn't get any objections so I'm going to merge this PR.
